File: kotlin-advent-of-code/src/main/kotlin/advent/twenty_fifteen/Day01NotQuiteLisp.kt
package advent.twenty_fifteen

import advent.common.DailyProblem

class Day01NotQuiteLisp : DailyProblem<Int, Int> {
    override var part1Answer: Int = 0
    override var part2Answer: Int = -1

    constructor (filename: String) {
        val input = Day01NotQuiteLisp::class.java.getResource(filename).readText()

        var floor = 0
        input.forEachIndexed { idx, c ->
            floor += when (c) {
                '(' -> +1
                ')' -> -1
                else -> 0
            }
            if (floor == -1 && part2Answer == -1) {
                part2Answer = idx + 1
            }
        }
        part1Answer = floor
    }
}

File: scala-advent-of-code/src/main/scala/advent/twenty_fifteen/Day01NotQuiteLisp.scala
package advent.twenty_fifteen

import advent.common.DailyProblem
import advent.utilities.FileUtilities

import scala.annotation.tailrec

class Day01NotQuiteLisp(filename: String) extends DailyProblem[Int, Int] {
  private val line = FileUtilities.readFile(filename).head

  @tailrec
  private def walk(line: List[Char], floor: Int, basement: Int, index: Int): (Int, Int) = {
    if (line.isEmpty) {
      (floor, basement)
    } else {
      val newFloor = if (line.head == '(') floor + 1 else floor - 1
      walk(line.tail, newFloor, if (newFloor == -1 && basement == -1) index else basement, index + 1)
    }
  }

  private val (finalFloor, firstBasement) = walk(line.toList, 0, -1, 1)

  override val part1Answer: Int = finalFloor
  override val part2Answer: Int = firstBasement
}
